test/charms/main/lib/charm.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
+
+import os
+import base64
+import pickle
+
+from juju.framework import (
+    EventBase,
+    Event,
+)
+
+from juju.charm import (
+    CharmBase,
+    CharmEvents,
+)
+
+import logging
+
+logger = logging.getLogger()
+
+
+class DBRelationJoinedEvent(EventBase):
+    pass
+
+
+class DBRelationChangedEvent(EventBase):
+    pass
+
+
+class DBRelationDepartedEvent(EventBase):
+    pass
+
+
+class DBRelationBrokenEvent(EventBase):
+    pass
+
+
+class MonRelationJoinedEvent(EventBase):
+    pass
+
+
+class MonRelationChangedEvent(EventBase):
+    pass
+
+
+class MonRelationDepartedEvent(EventBase):
+    pass
+
+
+class MonRelationBrokenEvent(EventBase):
+    pass
+
+
+class HARelationJoinedEvent(EventBase):
+    pass
+
+
+class HARelationChangedEvent(EventBase):
+    pass
+
+
+class HARelationDepartedEvent(EventBase):
+    pass
+
+
+class HARelationBrokenEvent(EventBase):
+    pass
+
+
+class MainCharmEvents(CharmEvents):
+    db_relation_joined = Event(DBRelationJoinedEvent)
+    db_relation_changed = Event(DBRelationChangedEvent)
+    db_relation_departed = Event(DBRelationDepartedEvent)
+    db_relation_broken = Event(DBRelationBrokenEvent)
+    mon_relation_joined = Event(MonRelationJoinedEvent)
+    mon_relation_changed = Event(MonRelationChangedEvent)
+    mon_relation_departed = Event(MonRelationDepartedEvent)
+    mon_relation_broken = Event(MonRelationBrokenEvent)
+    ha_relation_joined = Event(HARelationJoinedEvent)
+    ha_relation_changed = Event(HARelationChangedEvent)
+    ha_relation_departed = Event(HARelationDepartedEvent)
+    ha_relation_broken = Event(HARelationBrokenEvent)
+
+
+class Charm(CharmBase):
+
+    on = MainCharmEvents()
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        # Load attributes that might have been passed via an environment
+        # variable (used for altering charm behavior during testing).
+        charm_config = os.environ.get('CHARM_CONFIG')
+        if charm_config is not None:
+            self._charm_config = pickle.loads(base64.b64decode(charm_config))
+        else:
+            self._charm_config = {}
+
+        self._state_file = self._charm_config.get('STATE_FILE')
+        self._state = {}
+        self._state['event_type_names_observed'] = list()
+
+        # __init__ does not use reset_type because the goal is to
+        # avoid those values being reset on every object instantiation.
+
+        charm_events = dict()
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(self.on).__dict__.items()
+                             if type(instance) == Event})
```

Comment:
And then, why is this being done here?  The point of this test (as I understand it) is ensuring that the entrypoint logic is working correctly for a charm. But instead, this is super specific to the current implementation we have, and not likely to be seen in a normal charm. Instead, I'd expect this to be a much more normal charm, with proper lines that read like
```
framework.observe(self.on.some_event, self.on_some_event)
``` 
For two or three events which are different from each other in some fundamental way (one of them is install, another is a created symlink, another is dynamic, and that's about it?).